src/hooks/useSigningKeys.ts

```diff
@@ -0,0 +1,86 @@
+import { ethers } from "ethers";
+import { useState, useEffect, useContext } from "react";
+import { ErrorContext } from "common/context/ErrorContext";
+
+import { recoverPublicKey } from "features/vote/helpers/recoverPublicKey";
+import { derivePrivateKey } from "features/vote/helpers/derivePrivateKey";
+import usePrevious from "common/hooks/usePrevious";
+
+// User has to sign once per round.
+// public and private signing keys is indexed by address, and we ask them to resign if the roundMessage changes.
+export interface SigningKeys {
+  [address: string]: {
+    publicKey: string;
+    privateKey: string;
+    roundMessage: string;
+  };
+}
+
+export default function useSigningKeys(
+  signer: ethers.Signer | null,
+  address: string | null,
+  roundId: string
+) {
+  const [signingKeys, setSigningKeys] = useState<SigningKeys>(() => {
+    const storageKeys = localStorage.getItem("signingKeys");
+    return storageKeys ? (JSON.parse(storageKeys) as SigningKeys) : {};
+  });
+
+  const { addError } = useContext(ErrorContext);
+
+  const previousSigner = usePrevious(signer);
+  const previousAddress = usePrevious(address);
+
+  useEffect(() => {
+    if (
+      signer &&
+      address &&
+      roundId &&
+      (previousAddress === null || previousSigner === null)
```

Comment:
Can you explain what this does? Is it possible this could block updates if the user swaps accounts?

Reply:
Basically because of the way we use react-query, these values update in different render cycles. I believe this catch is in place to prevent the user having to sign twice.